### PR TITLE
[KOGITO-8889] Supporting expression in array items

### PR DESCRIPTION
**Pull request by fjtirado**
As described in the JIRA, array items were not evaluated as expressions.

### CODE REVIEW

File: kogito-serverless-workflow/kogito-serverless-workflow-executor/src/test/java/org/kie/kogito/serverless/workflow/executor/StaticFluentWorkflowApplicationTest.java [hunk 1]
```diff
@@ -56,19 +58,20 @@ void helloWorld() {
         }
     }
 
+    @RegisterExtension
```
Comment (fjtirado) [inline kogito-serverless-workflow/kogito-serverless-workflow-executor/src/test/java/org/kie/kogito/serverless/workflow/executor/StaticFluentWorkflowApplicationTest.java:61]: This change is not strictly related with the PR but this test was using always 8080, which surprinsingly does not create any conflict (probably because other test are well written, I called this the all except me are vaccinated ;))

File: kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/parser/types/ServiceTypeHandler.java [hunk 2]
```diff
@@ -60,7 +60,7 @@ protected WorkItemNodeFactory<?> buildWorkItem(RuleFlowNodeContainerFactory<?, ?
     @Override
     protected <T extends RuleFlowNodeContainerFactory<T, ?>> WorkItemNodeFactory<T> addFunctionArgs(Workflow workflow, WorkItemNodeFactory<T> node, FunctionRef functionRef) {
         JsonNode functionArgs = functionRef.getArguments();
-        if (functionArgs == null || functionArgs.isEmpty()) {
+        if (functionArgs == null) {
```
Comment (fjtirado) [inline kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/parser/types/ServiceTypeHandler.java:63]: isEmpty return true for primitives, so we are only passing the whole model to the service (legacy functionality using in the service example) if there are not args (args are null)

File: kogito-workitems/kogito-jackson-utils/src/main/java/org/kie/kogito/jackson/utils/JsonObjectUtils.java [hunk 3]
```diff
@@ -146,13 +156,9 @@ private static Object internalToJavaValue(JsonNode jsonNode, Function<JsonNode,
         } else if (jsonNode.isNumber()) {
             return jsonNode.numberValue();
         } else if (jsonNode.isArray()) {
-            Collection result = new ArrayList<>();
-            for (JsonNode item : ((ArrayNode) jsonNode)) {
-                result.add(internalToJavaValue(item, function));
-            }
-            return result;
+            return arrayFunction.apply((ArrayNode) jsonNode);
```
Comment (fjtirado) [inline kogito-workitems/kogito-jackson-utils/src/main/java/org/kie/kogito/jackson/utils/JsonObjectUtils.java:159]: This is the key change to make arrays be evaluated, if simpleToJavaValue is invoked (the method used during build) the arraynode is not converted to list (that we leave for toJavaValue method during runtime)

File: kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/utils/WorkItemBuilder.java [hunk 4]
```diff
@@ -75,11 +79,13 @@ protected final void processArgs(Workflow workflow, WorkItemNodeFactory<?> workI
...
         } else {
             Object object = functionReference(workflow, JsonObjectUtils.simpleToJavaValue(functionArgs));
-            if (isExpression(workflow, object)) {
+            boolean isExpr = isExpression(workflow, object);
+            if (isExpr) {
                 workItemFactory.workParameterFactory(new ExpressionParametersFactorySupplier(workflow.getExpressionLang(), object, paramName));
             } else {
                 workItemFactory.workParameter(RestWorkItemHandler.CONTENT_DATA, object);
             }
+            workItemFactory.workParameterDefinition(RestWorkItemHandler.CONTENT_DATA, getDataType(object, isExpr));
```
Comment (fjtirado) [inline kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/utils/WorkItemBuilder.java:88]: This is needed to make service calls work when there is only one parameter (or directly an array not embedded into an argument)

File: kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/utils/WorkItemBuilder.java [hunk 5]
```diff
@@ -111,7 +117,17 @@ private void processArg(Workflow workflow, String key, Object value, WorkItemNod
                 .workParameter(key,
                         isExpr ? new ObjectResolverSupplier(workflow.getExpressionLang(), value, paramName) : value)
                 .workParameterDefinition(key,
-                        DataTypeResolver.fromObject(value, isExpr));
+                        getDataType(value, isExpr));
+    }
+
+    DataType getDataType(Object object, boolean isExpr) {
+        if (object instanceof ObjectNode) {
```
Comment (fjtirado) [inline kogito-serverless-workflow/kogito-serverless-workflow-builder/src/main/java/org/kie/kogito/serverless/workflow/utils/WorkItemBuilder.java:124]: This is telling the work item (this is used by service calls) that ObjectNodes are converted to Map and ArrayNodes to Collections by JsonObjectUtils.toJavaValue

File: kogito-serverless-workflow/kogito-serverless-workflow-runtime/src/main/java/org/kie/kogito/serverless/workflow/workitemparams/ObjectResolver.java [hunk 6]
```diff
@@ -26,6 +26,6 @@ public ObjectResolver(String exprLang, Object expr, String paramName) {
 
     @Override
     public Object apply(KogitoWorkItem workItem) {
-        return JsonObjectUtils.simpleToJavaValue(evalExpression(workItem));
+        return JsonObjectUtils.toJavaValue(evalExpression(workItem));
```
Comment (fjtirado) [inline kogito-serverless-workflow/kogito-serverless-workflow-runtime/src/main/java/org/kie/kogito/serverless/workflow/workitemparams/ObjectResolver.java:29]: This change aligns ObjectResolver with ExpressionParametersFactory

File: kogito-serverless-workflow/kogito-serverless-workflow-executor/src/test/java/org/kie/kogito/serverless/workflow/executor/StaticFluentWorkflowApplicationTest.java [hunk 7]
```diff
@@ -56,19 +58,20 @@ void helloWorld() {
...
+    static WireMockExtension wm1 = WireMockExtension.newInstance()
```
Comment (hbelmiro) [inline kogito-serverless-workflow/kogito-serverless-workflow-executor/src/test/java/org/kie/kogito/serverless/workflow/executor/StaticFluentWorkflowApplicationTest.java:62]:
```suggestion
    static WireMockExtension wm = WireMockExtension.newInstance()
```

File: kogito-serverless-workflow/kogito-serverless-workflow-executor/src/test/java/org/kie/kogito/serverless/workflow/executor/StaticFluentWorkflowApplicationTest.java [hunk 8]
```diff
@@ -56,19 +58,20 @@ void helloWorld() {
...
+            .options(wireMockConfig().dynamicPort())
+            .build();
+
     @Test
     void restInvocation() {
-        WireMockServer server = new WireMockServer();
         JsonNode expectedOutput = ObjectMapperFactory.get().createObjectNode().put("name", "Javierito");
-        server.stubFor(get("/name").willReturn(aResponse().withStatus(200).withJsonBody(expectedOutput)));
+        wm1.stubFor(get("/name").willReturn(aResponse().withStatus(200).withJsonBody(expectedOutput)));
```
Comment (hbelmiro) [inline kogito-serverless-workflow/kogito-serverless-workflow-executor/src/test/java/org/kie/kogito/serverless/workflow/executor/StaticFluentWorkflowApplicationTest.java:69]:
```suggestion
        wm.stubFor(get("/name").willReturn(aResponse().withStatus(200).withJsonBody(expectedOutput)));
```

File: kogito-serverless-workflow/kogito-serverless-workflow-executor/src/test/java/org/kie/kogito/serverless/workflow/executor/StaticFluentWorkflowApplicationTest.java [hunk 9]
```diff
@@ -56,19 +58,20 @@ void helloWorld() {
...
         final String FUNCTION_NAME = "function";
         try (StaticWorkflowApplication application = StaticWorkflowApplication.create()) {
-            server.start();
-            Workflow workflow = workflow("HelloRest").function(rest(FUNCTION_NAME, HttpMethod.get, "http://localhost:" + server.getOptions().portNumber() + "/name"))
+            Workflow workflow = workflow("HelloRest").function(rest(FUNCTION_NAME, HttpMethod.get, "http://localhost:" + wm1.getPort() + "/name"))
```
Comment (hbelmiro) [inline kogito-serverless-workflow/kogito-serverless-workflow-executor/src/test/java/org/kie/kogito/serverless/workflow/executor/StaticFluentWorkflowApplicationTest.java:72]:
```suggestion
            Workflow workflow = workflow("HelloRest").function(rest(FUNCTION_NAME, HttpMethod.get, "http://localhost:" + wm.getPort() + "/name"))
```

**Review by hbelmiro — APPROVED** [review]
LGTM.
I just left a comment regarding a variable name.

**Review by cristianonicolai — APPROVED** [review]
